code/move/solution/trymove.sol.cpp

```diff
@@ -1,39 +1,87 @@
-#include <iostream>
-#include <vector>
+#include <memory>
 #include <random>
-#include "NVector.sol.hpp"
-
-template<typename T>
-void swap(std::vector<T> &v, int a, int b) {
-    T mem = std::move(v[a]);
-    v[a] = std::move(v[b]);
-    v[b] = std::move(mem);
-    // even better
-    //swap(v[a], v[b]);
-}
+#include <vector>
 
-template<typename T>
-void randomize(std::vector<T> &v) {
-    // we randomize via len random inversions
-    std::size_t len = v.size();
-    std::default_random_engine e;
-    std::uniform_int_distribution<std::size_t> d{0, len};
-    for (std::size_t i = 0; i < len; i++) {
-        int a = d(e);
-        int b = d(e);
-        swap(v, a, b);
+/**
+ * An attempt to make a movable class. It allocates memory on construction, and has
+ * array semantics. (Prefer std::array or std::vector instead of hand-crafted solutions, though.)
+ * We made it copyable by providing copy constructor and assignment operator, but note that we
+ * are violating the rule of 5, so it's not moveable yet.
+*/
+class Moveable {
+    static constexpr std::size_t size = 10000;
+
+public:
+    Moveable() = default;
+    Moveable(Moveable const & other) :
+        Moveable()
+    {
+        // Delegate copying to the assignment operator
+        *this = other;
     }
-}
+    Moveable & operator=(const Moveable & other)
+    {
+        if (other.m_storage != m_storage) {
+            std::copy(other.m_storage.get(), other.m_storage.get() + size, m_storage.get());
+        }
+        return *this;
+    }
+
+    // *******
+    // Task 1: Provide move constructor and move assignment operator
+    // *******
+
+    // Note how we are avoiding the initial allocation by directly
+    // move-initialising the pointer
+    Moveable(Moveable && other) noexcept :
+        m_storage{ std::move(other.m_storage) }
+    {  }
+    Moveable & operator=(Moveable && other) noexcept
+    {
+        if (m_storage != other.m_storage) m_storage = std::move(other.m_storage);
+        return *this;
+    }
+
+    int & operator[](std::size_t index) { return m_storage[index]; }
+
+private:
+    std::unique_ptr<int[]> m_storage { new int[size] };
```

Comment:
For me, naked `new` is a code smell, so I would use here:
```suggestion
    std::unique_ptr<int[]> m_storage = std::make_unique<int[]>(size);
```

Reply:
I would have to upgrade to c++17 to do this. I wasn't sure if that should be an option.

Reply:
C++17 is the minimum for all exercises already, see e.g. here: https://github.com/hsf-training/cpluspluscourse/pull/356#issuecomment-1276706146